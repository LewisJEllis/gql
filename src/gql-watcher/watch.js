/* @flow */
import os from 'os';
import { type FileMatchConfig } from 'gql-config/types';
import sane from 'sane';
import globby from 'globby';
import _debounce from 'lodash/debounce';
import parseFileMatchConfig from 'gql-shared/parseFileMatchConfig';

export type WatchFile = $Exact<{
  name: string,
  exists: boolean,
}>;

class WatchEventsBatcher {
  _listener: (files: Array<WatchFile>) => void;
  _queue = [];
  _activeTimeout = null;
  _waitTimeForBatching = 200 /* msec */;

  constructor(listener: (files: Array<WatchFile>) => void) {
    this._listener = listener;
  }

  __dispatchQueue = _debounce(() => {
    const events = this._queue;
    this._queue = [];

    if (events.length === 0) {
      return;
    }

    this._listener(events);
  }, this._waitTimeForBatching);

  add(eventOrEvents: WatchFile | Array<WatchFile>) {
    this._queue.push(
      ...(Array.isArray(eventOrEvents) ? eventOrEvents : [eventOrEvents]),
    );
    this.__dispatchQueue();
  }
}

export type WatchOptions = {
  rootPath: string,
  files: FileMatchConfig,
  name: string,
  watchman?: boolean,
  watch?: boolean,
  onChange: (changedFiles: Array<WatchFile>) => void,
};

function boolValue(value: ?boolean, defaultValue: boolean) {
  return typeof value === 'boolean' ? value : defaultValue;
}

function setupWatcher({ rootPath, onChange, name, glob, ignored, watchman }) {
  const watchEventsBatcher = new WatchEventsBatcher(onChange);
  const useWatchman = boolValue(watchman, true);

  /* eslint-disable no-nested-ternary */
  const WatcherClass = useWatchman
    ? sane.WatchmanWatcher
    : os.platform() === 'darwin' ? sane.FSEventWatcher : sane.NodeWatcher;
  /* eslint-disable ternary */

  const watcher = new WatcherClass(rootPath, {
    glob,
    ignored,
  });

  // console.log(`[${WatcherClass.name}] '${name}' starting...`);
  const key = `[${WatcherClass.name}] '${name}' started`;
  // console.time(key);
  watcher.on('ready', () => {
    // console.timeEnd(key);
  });

  watcher.on('change', filePath => {
    watchEventsBatcher.add({ name: filePath, exists: true });
  });

  watcher.on('add', filePath => {
    watchEventsBatcher.add({ name: filePath, exists: true });
  });

  watcher.on('delete', filePath => {
    watchEventsBatcher.add({ name: filePath, exists: false });
  });

  return watcher;
}

export default function watch(options: WatchOptions) {
  const { glob, ignored } = parseFileMatchConfig(options.files);

  let watcher = null;
  const onReadyPromise = globby(glob, { cwd: options.rootPath })
    .then(matches => {
      const watchFiles = matches.map(file => ({ name: file, exists: true }));
      if (!boolValue(options.watch, true)) {
        return watchFiles;
      }
      // if in watchmode then wait for watcher to start
      return new Promise(resolve => {
        watcher = setupWatcher({
          rootPath: options.rootPath,
          onChange: options.onChange,
          watchman: options.watchman,
          name: options.name,
          glob,
          ignored,
        });
        watcher.on('ready', resolve);
      }).then(() => watchFiles);
    })
    .then(watchFiles => {
      options.onChange(watchFiles);
    })
    .catch(err => {
      throw err;
    });

  return {
    onReady() {
      return onReadyPromise;
    },
    close() {
      if (watcher) {
        watcher.close();
      }
    },
  };
}
