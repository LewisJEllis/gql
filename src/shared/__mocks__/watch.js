/* @flow */
import glob from 'glob';

export default function watch({ rootPath, files, onChange }: any) {
  watch.__files = glob.sync(files, { cwd: rootPath }).map(_name => ({ name: _name, exists: true }));
  watch.__onChangeCallback = onChange;
  return { end: () => {} };
}

watch.__triggerChange = (files) => {
  watch.__onChangeCallback(files || watch.__files);
};
