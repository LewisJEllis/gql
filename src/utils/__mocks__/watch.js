/* @flow */
import glob from 'glob';

export default function watch(rootPath: string, pattern: string, callback: Function) {
  watch.__files = glob.sync(pattern, { cwd: rootPath }).map(name => ({ name, exists: true }));
  watch.__onChangeCallback = callback;
}

watch.__triggerChange = (files) => {
  watch.__onChangeCallback(files || watch.__files);
};
