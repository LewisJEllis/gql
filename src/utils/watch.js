/* @flow */
import watchman from 'fb-watchman';
import type { WatchFile } from './types';

const client = new watchman.Client();

export default function watch(
  rootPath: string,
  glob: string,
  callback: (changedFiles: Array<WatchFile>) => void,
) {
  console.log(`Launching watch server for ${rootPath}/${glob}`);
  client.capabilityCheck({ optional: [], required: ['relative_root'] }, (error) => {
    if (error) {
      console.error(error);
      client.end();
      return;
    }

    client.command(['watch-project', rootPath], (err, { watch: _watch, relative_path: relativePath, warning }) => {
      if (err) {
        console.error('Error initiating watch:', err);
        return;
      }

      if (warning) {
        console.log('warning: ', warning);
      }

      const sub = {
        expression: ['allof', ['match', glob, 'wholename']],
        fields: ['name', 'exists', 'type'],
        relative_root: undefined,
      };

      if (relativePath) {
        sub.relative_root = relativePath;
      }

      console.log(`Watch established on ${_watch} relative_path ${relativePath}`);

      const gqlFilesWatchSubscription = 'gqlFilesWatch';
      client.command(['subscribe', _watch, gqlFilesWatchSubscription, sub], (subscribeError, resp) => {
        if (subscribeError) {
          console.error('failed to subscribe: ', subscribeError);
          return;
        }

        console.log(`subsciption ${resp.subscribe} established`);
      });

      client.on('subscription', (resp) => {
        if (resp.subscription !== gqlFilesWatchSubscription) {
          return;
        }
        callback(resp.files);
      });
    });
  });
}

