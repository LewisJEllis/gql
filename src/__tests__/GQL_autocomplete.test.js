/* @flow */
import watch from '../utils/watch';
import { GQL } from '../GQL';
import path from 'path';

jest.mock('../utils/watch');

describe('autocomplete', () => {
  let gql;

  beforeEach(() => {
    gql = new GQL({ cwd: path.resolve('src/__test-data__/') });
    // $FlowDisableNextLine
    watch.__triggerChange();
  });

  describe('autocomplete for `type`', () => {
    it('show all types for Field', () => {
      const sourceText = `
        type TestType {
          name:   #
        }
      `;
      const position = { line: 3, column: 15 };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });

    it('show all types for Field even after typing few chars', () => {
      const sourceText = `
        type TestType {
          name: S #
        }
      `;
      const position = { line: 3, column: 17 };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });

    it('show all output types for Field if cursor on ! and of the form T!', () => {
      const sourceText = `
        type TestType {
          name: S! #
        }
      `;
      const position = { line: 3, column: 18 /* cursor on ! */ };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });

    it('show all output types for Field if cursor on ] and of type [T]', () => {
      const sourceText = `
        type TestType {
          name: [S] #
        }
      `;
      const position = { line: 3, column: 19 /* cursor on ! */ };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });

    it('show all output types for Field if cursor on ! and of the form [T!]', () => {
      const sourceText = `
        type TestType {
          name: [S!] #
        }
      `;
      const position = { line: 3, column: 19 /* cursor on ! */ };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });

    it('dont show any types for Field if cursor after !', () => {
      const sourceText = `
        type TestType {
          name: S! #
        }
      `;
      const position = { line: 3, column: 19 /* cursor after ! */ };
      expect(gql.autocomplete(sourceText, position)).toEqual([]);
    });

    it('dont show any types for Field if cursor after !', () => {
      const sourceText = `
        type TestType {
          name: S! #
        }
      `;
      const position = { line: 3, column: 19 /* cursor after ! */ };
      expect(gql.autocomplete(sourceText, position)).toEqual([]);
    });

    it('show all types for Field with args', () => {
      const sourceText = `
        type TestType {
          players(first: String!):  #
        }
      `;
      const position = { line: 3, column: 36 };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });

    it('show all types for Field ListType', () => {
      const sourceText = `
        type TestType {
          name: []
        }
      `;
      const position = { line: 3, column: 18 };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });

    it('show only InputType for Field input args ', () => {
      const sourceText = `
        type TestType {
          name(xyz: )
        }
      `;
      const position = { line: 3, column: 20 };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });

    it('show `implements` && `{` after typename', () => {
      const sourceText = `
        type TestType  #
      `;

      const position = { line: 2, column: 23 };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });

    it('show all `GraphQLInterfaceType` types after `implements` keyword', () => {
      const sourceText = `
        type TestType implements  #
      `;
      const position = { line: 2, column: 34 };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });

    it('show all `GraphQLInterfaceType` types after multiple InterfaceTypes', () => {
      const sourceText = `
        type TestType implements Node  #
      `;
      const position = { line: 2, column: 39 };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });
  });

  describe('autocomplete for `input`', () => {
    it('show only InputType for field', () => {
      const sourceText = `
        input TestTypeInput {
          name:    #
        }
      `;
      const position = { line: 3, column: 15 };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });

    it('show only InputType for field of ListType', () => {
      const sourceText = `
        input TestType {
          name: []
        }
      `;
      const position = { line: 3, column: 18 };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });
  });

  describe('autocomplete for `union`', () => {
    it('show only GrahQLObjectType for field', () => {
      const sourceText = `
        union TestUnionType =  #
      `;
      const position = { line: 2, column: 31 };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });

    it('show `|` after Type(unionMember)', () => {
      const sourceText = `
        union TestUnionType =  TestType  #
      `;
      const position = { line: 2, column: 41 };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });

    it('show all GrahQLObjectType after `|`', () => {
      const sourceText = `
        union TestUnionType =  TestType |  #
      `;
      const position = { line: 2, column: 43 };
      expect(gql.autocomplete(sourceText, position)).toMatchSnapshot();
    });
  });
});
