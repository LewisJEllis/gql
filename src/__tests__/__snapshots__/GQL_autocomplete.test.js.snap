exports[`autocomplete autocomplete for \`input\` show only InputType for field 1`] = `
Array [
  Object {
    "description": "",
    "text": "PlayerCreateInput",
    "type": "GraphQLInputObjectType",
  },
  Object {
    "description": "",
    "text": "CustomScalar",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.",
    "text": "String",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Int\` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ",
    "text": "Int",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Float\` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ",
    "text": "Float",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Boolean\` scalar type represents \`true\` or \`false\`.",
    "text": "Boolean",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`"4"\`) or integer (such as \`4\`) input value will be accepted as an ID.",
    "text": "ID",
    "type": "GraphQLScalarType",
  },
]
`;

exports[`autocomplete autocomplete for \`input\` show only InputType for field of ListType 1`] = `
Array [
  Object {
    "description": "",
    "text": "PlayerCreateInput",
    "type": "GraphQLInputObjectType",
  },
  Object {
    "description": "",
    "text": "CustomScalar",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.",
    "text": "String",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Int\` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ",
    "text": "Int",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Float\` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ",
    "text": "Float",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Boolean\` scalar type represents \`true\` or \`false\`.",
    "text": "Boolean",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`"4"\`) or integer (such as \`4\`) input value will be accepted as an ID.",
    "text": "ID",
    "type": "GraphQLScalarType",
  },
]
`;

exports[`autocomplete autocomplete for \`type\` show \`implements\` && \`{\` after typename 1`] = `
Array [
  Object {
    "text": "implements",
    "type": "Implements",
  },
  Object {
    "text": "{",
  },
]
`;

exports[`autocomplete autocomplete for \`type\` show all \`GraphQLInterfaceType\` types after \`implements\` keyword 1`] = `
Array [
  Object {
    "description": "",
    "text": "Node",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "Edge",
    "type": "GraphQLInterfaceType",
  },
]
`;

exports[`autocomplete autocomplete for \`type\` show all \`GraphQLInterfaceType\` types after multiple InterfaceTypes 1`] = `
Array [
  Object {
    "description": "",
    "text": "Node",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "Edge",
    "type": "GraphQLInterfaceType",
  },
]
`;

exports[`autocomplete autocomplete for \`type\` show all output types for Field if cursor on ! and of the form [T!] 1`] = `
Array [
  Object {
    "description": "Mutation contains all allowed mutations",
    "text": "Mutation",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "PlayerCreatePayload",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "Query is the root query object",
    "text": "Query",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Node",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "Player",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Team",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Viewer",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Edge",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "CustomScalar",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.",
    "text": "String",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Int\` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ",
    "text": "Int",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Float\` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ",
    "text": "Float",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Boolean\` scalar type represents \`true\` or \`false\`.",
    "text": "Boolean",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`\"4\"\`) or integer (such as \`4\`) input value will be accepted as an ID.",
    "text": "ID",
    "type": "GraphQLScalarType",
  },
]
`;

exports[`autocomplete autocomplete for \`type\` show all output types for Field if cursor on ! and of the form T! 1`] = `
Array [
  Object {
    "description": "Mutation contains all allowed mutations",
    "text": "Mutation",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "PlayerCreatePayload",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "Query is the root query object",
    "text": "Query",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Node",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "Player",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Team",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Viewer",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Edge",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "CustomScalar",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.",
    "text": "String",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Int\` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ",
    "text": "Int",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Float\` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ",
    "text": "Float",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Boolean\` scalar type represents \`true\` or \`false\`.",
    "text": "Boolean",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`\"4\"\`) or integer (such as \`4\`) input value will be accepted as an ID.",
    "text": "ID",
    "type": "GraphQLScalarType",
  },
]
`;

exports[`autocomplete autocomplete for \`type\` show all output types for Field if cursor on ] and of type [T] 1`] = `
Array [
  Object {
    "description": "Mutation contains all allowed mutations",
    "text": "Mutation",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "PlayerCreatePayload",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "Query is the root query object",
    "text": "Query",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Node",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "Player",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Team",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Viewer",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Edge",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "CustomScalar",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.",
    "text": "String",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Int\` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ",
    "text": "Int",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Float\` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ",
    "text": "Float",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Boolean\` scalar type represents \`true\` or \`false\`.",
    "text": "Boolean",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`\"4\"\`) or integer (such as \`4\`) input value will be accepted as an ID.",
    "text": "ID",
    "type": "GraphQLScalarType",
  },
]
`;

exports[`autocomplete autocomplete for \`type\` show all types for Field 1`] = `
Array [
  Object {
    "description": "Mutation contains all allowed mutations",
    "text": "Mutation",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "PlayerCreatePayload",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "Query is the root query object",
    "text": "Query",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Node",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "Player",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Team",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Viewer",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Edge",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "CustomScalar",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.",
    "text": "String",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Int\` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ",
    "text": "Int",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Float\` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ",
    "text": "Float",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Boolean\` scalar type represents \`true\` or \`false\`.",
    "text": "Boolean",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`"4"\`) or integer (such as \`4\`) input value will be accepted as an ID.",
    "text": "ID",
    "type": "GraphQLScalarType",
  },
]
`;

exports[`autocomplete autocomplete for \`type\` show all types for Field ListType 1`] = `
Array [
  Object {
    "description": "Mutation contains all allowed mutations",
    "text": "Mutation",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "PlayerCreatePayload",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "Query is the root query object",
    "text": "Query",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Node",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "Player",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Team",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Viewer",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Edge",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "CustomScalar",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.",
    "text": "String",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Int\` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ",
    "text": "Int",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Float\` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ",
    "text": "Float",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Boolean\` scalar type represents \`true\` or \`false\`.",
    "text": "Boolean",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`"4"\`) or integer (such as \`4\`) input value will be accepted as an ID.",
    "text": "ID",
    "type": "GraphQLScalarType",
  },
]
`;

exports[`autocomplete autocomplete for \`type\` show all types for Field even after typing few chars 1`] = `
Array [
  Object {
    "description": "Mutation contains all allowed mutations",
    "text": "Mutation",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "PlayerCreatePayload",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "Query is the root query object",
    "text": "Query",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Node",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "Player",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Team",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Viewer",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Edge",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "CustomScalar",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.",
    "text": "String",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Int\` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ",
    "text": "Int",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Float\` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ",
    "text": "Float",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Boolean\` scalar type represents \`true\` or \`false\`.",
    "text": "Boolean",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`\"4\"\`) or integer (such as \`4\`) input value will be accepted as an ID.",
    "text": "ID",
    "type": "GraphQLScalarType",
  },
]
`;

exports[`autocomplete autocomplete for \`type\` show all types for Field with args 1`] = `
Array [
  Object {
    "description": "Mutation contains all allowed mutations",
    "text": "Mutation",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "PlayerCreatePayload",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "Query is the root query object",
    "text": "Query",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Node",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "Player",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Team",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Viewer",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Edge",
    "type": "GraphQLInterfaceType",
  },
  Object {
    "description": "",
    "text": "CustomScalar",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.",
    "text": "String",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Int\` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ",
    "text": "Int",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Float\` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ",
    "text": "Float",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Boolean\` scalar type represents \`true\` or \`false\`.",
    "text": "Boolean",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`"4"\`) or integer (such as \`4\`) input value will be accepted as an ID.",
    "text": "ID",
    "type": "GraphQLScalarType",
  },
]
`;

exports[`autocomplete autocomplete for \`type\` show only InputType for Field input args  1`] = `
Array [
  Object {
    "description": "",
    "text": "PlayerCreateInput",
    "type": "GraphQLInputObjectType",
  },
  Object {
    "description": "",
    "text": "CustomScalar",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.",
    "text": "String",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Int\` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ",
    "text": "Int",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Float\` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ",
    "text": "Float",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`Boolean\` scalar type represents \`true\` or \`false\`.",
    "text": "Boolean",
    "type": "GraphQLScalarType",
  },
  Object {
    "description": "The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`"4"\`) or integer (such as \`4\`) input value will be accepted as an ID.",
    "text": "ID",
    "type": "GraphQLScalarType",
  },
]
`;

exports[`autocomplete autocomplete for \`union\` show \`|\` after Type(unionMember) 1`] = `
Array [
  Object {
    "description": "OR symbol",
    "text": "|",
    "type": "OR",
  },
]
`;

exports[`autocomplete autocomplete for \`union\` show all GrahQLObjectType after \`|\` 1`] = `
Array [
  Object {
    "description": "Mutation contains all allowed mutations",
    "text": "Mutation",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "PlayerCreatePayload",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "Query is the root query object",
    "text": "Query",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Player",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Team",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Viewer",
    "type": "GraphQLObjectType",
  },
]
`;

exports[`autocomplete autocomplete for \`union\` show only GrahQLObjectType for field 1`] = `
Array [
  Object {
    "description": "Mutation contains all allowed mutations",
    "text": "Mutation",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "PlayerCreatePayload",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "Query is the root query object",
    "text": "Query",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Player",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Team",
    "type": "GraphQLObjectType",
  },
  Object {
    "description": "",
    "text": "Viewer",
    "type": "GraphQLObjectType",
  },
]
`;
